### PR TITLE
Add Extended Isolation Forest (EIF) support to the Scala/Spark isolation-forest library.

### PR DESCRIPTION
**Pull request by jverbus**
## Summary

This PR adds **Extended Isolation Forest (EIF)** support to the Scala/Spark `isolation-forest` library.

The new implementation introduces a Spark ML `Estimator`/`Model` pair for EIF, adds save/load support for the extended tree format, expands test coverage substantially, and updates the README with usage examples, parameter documentation, benchmarks, and clear notes about ONNX support.

In addition to the new EIF functionality, this PR also refactors shared training/persistence utilities and improves robustness and backward compatibility for the existing standard `IsolationForest` implementation.

## Background

The existing library only supported the standard Isolation Forest algorithm, which uses **axis-aligned splits**. That works well in many cases, but it can introduce directional bias and may underperform when anomalies lie along **correlated** or **non-axis-aligned** directions.

Extended Isolation Forest addresses this by replacing axis-aligned splits with **random hyperplane splits**, making the detector more rotationally invariant and better suited for data with correlated features.

This PR implements EIF based on the original paper and validates behavior against the reference implementation:

- Original EIF paper: [Hariri et al., "Extended Isolation Forest"](https://arxiv.org/abs/1811.02141)
- Reference implementation: [sahandha/eif](https://github.com/sahandha/eif)

### 1. New Extended Isolation Forest implementation

Added a new package:

- `com.linkedin.relevance.isolationforest.extended`

with the following new public Spark ML APIs:

- `ExtendedIsolationForest`
- `ExtendedIsolationForestModel`

and the supporting implementation types:

- `ExtendedIsolationForestParams`
- `ExtendedIsolationTree`
- `ExtendedNodes`
- `ExtendedUtils`
- `ExtendedIsolationForestModelReadWrite`

### 2. Random hyperplane splits for EIF

Each EIF tree now isolates points using **random hyperplanes** instead of single-feature thresholds.

Implementation details:

- hyperplane normals are sampled randomly and **L2-normalized**
- hyperplanes are stored **sparsely**
- only the non-zero coordinates are persisted and used at score time
- coordinates are canonicalized into sorted order for stable persistence/debugging

### 3. `extensionLevel` parameter

EIF introduces a new parameter:

- `extensionLevel`

Behavior:

- `0` -> axis-aligned EIF splits
- `numFeatures - 1` -> fully extended hyperplanes
- if unset, it defaults at fit time to the fully extended setting for the resolved feature subspace

Important semantics captured in the implementation and docs:

- `extensionLevel` is interpreted relative to the **resolved per-tree feature subspace**
- when `maxFeatures < 1.0`, the valid `extensionLevel` range is based on that subspace, not the original dataset dimension
- invalid values fail fast during fit

### 4. EIF model save/load support

Added full persistence support for `ExtendedIsolationForestModel`.

This includes:

- metadata read/write
- Avro serialization of extended trees
- reconstruction of tree structure on load
- persistence of sparse hyperplane data:
  - indices
  - weights
  - offset

A saved EIF model can now be loaded with:

```scala
val model = ExtendedIsolationForestModel.load(path)
```

---

## Standard Isolation Forest improvements included in this PR

While adding EIF, this PR also cleans up and strengthens the existing standard IF implementation.

### 1. Shared training/schema utilities

Refactored common logic into reusable helpers so both standard IF and EIF use the same code paths where appropriate.

New shared helpers include:

- schema validation/output column appending
- parameter resolution for `maxFeatures` / `maxSamples`
- sampled dataset creation
- tree training orchestration
- threshold computation

### 2. Shared read/write metadata utilities

Extracted common metadata handling into:

- `core/IsolationForestModelReadWriteUtils.scala`

This reduces duplication between standard IF and EIF persistence code.

### 3. Track `totalNumFeatures` in models

Standard `IsolationForestModel` now records:

- `numFeatures` = resolved per-tree feature count
- `totalNumFeatures` = full training input dimension

This enables feature-dimension validation during scoring and makes persisted model metadata more explicit.

### 4. Backward-compatible standard IF constructor behavior

To preserve compatibility, standard IF now has:

- an internal 5-argument constructor that carries `totalNumFeatures`
- a restored public 4-argument constructor that preserves the legacy unknown-dimension path

This keeps older construction patterns working while allowing newly trained or newly loaded models to retain full dimensionality metadata.

Related to that, `ExtendedIsolationForestModel` keeps its 5-argument constructor internal-only so EIF model construction remains package-scoped.

### 5. Legacy model loading support

Older saved standard IF models that do not contain `totalNumFeatures` metadata still load successfully.

In that case the model follows an **unknown-dimension** path and skips feature-size validation.

### 6. More robust scoring behavior

Standard and extended models now fail fast for invalid scoring scenarios such as:

- empty ensembles
- `numSamples < 2`
- feature vectors whose dimension does not match the model’s known training dimension

### 7. Small validation/robustness fixes

Additional guardrails added in shared training logic:

- resolved `maxSamples` must be at least `2`
- explicit failure if a training partition ends up with zero samples
- schema validation logic is centralized and reused

---

## EIF implementation details worth calling out

A few implementation choices are deliberate and important.

### Degenerate splits and zero-size leaves

EIF can legitimately produce degenerate hyperplane splits where all points land on one side.  
To match the EIF algorithm/reference behavior, this implementation:

- does **not** retry the split
- allows zero-size leaves via `ExtendedExternalNode(0)`

This is different from the standard IF behavior, where axis-aligned feature selection retries until a splittable feature is found.

### Sparse hyperplane representation

Hyperplanes are stored in the original feature space using only the non-zero coordinates.  
This keeps:

- node-local storage proportional to the number of active coordinates
- dot-product work proportional to the active coordinates, not the full input dimension

### `ExtendedIF_0` is not identical to standard IF

Although `extensionLevel = 0` yields axis-aligned hyperplanes, it is still **EIF behavior**, not standard IF behavior.

Key differences include:

- standard IF retries when a constant feature is selected
- EIF commits to the random draw
- RNG consumption patterns differ

This distinction is now documented in the README and discussed explicitly in the benchmark section.

---

## Documentation updates

The README was expanded significantly to reflect the new functionality.

### Added documentation for EIF

New sections include:

- Extended Isolation Forest overview
- when to use EIF vs standard IF
- `extensionLevel` parameter semantics
- Scala usage example
- save/load example

### Clarified ONNX support

The README now explicitly states:

- ONNX export is supported for **standard `IsolationForestModel` only**
- EIF is **not** currently supported by the ONNX converter because the converter assumes axis-aligned tree ensembles

### Benchmarks section updated

The README benchmark section now compares:

- `StandardIF`
- `ExtendedIF_0`
- `ExtendedIF_max`
- Liu et al. 2008 results
- the reference Python EIF implementation (`sahandha/eif`)

It also summarizes where EIF helps most and where it may underperform.

### Additional README cleanup

Also included:

- dependency examples updated to current artifact naming/version style
- version references updated from Spark `3.5.1` to `3.5.5`
- citation section added
- Scala/Python example cleanup and typo fixes

---

## Tests added / updated

This PR adds substantial new test coverage for both EIF and supporting standard IF changes.

### New EIF test suites

Added:

- `ExtendedIsolationForestTest`
- `ExtendedIsolationForestModelWriteReadTest`
- `ExtendedIsolationTreeTest`

Coverage includes:

- training/scoring on benchmark datasets
- zero contamination behavior
- default and explicit `extensionLevel`
- invalid `extensionLevel` handling
- persistence round-trip
- saved model structure snapshot tests
- identical-feature edge cases
- zero-size leaf handling
- sparse hyperplane validation
- L2 normalization of hyperplane normals
- feature-dimension validation
- empty model / invalid model transform failures

### Updated standard IF tests

Expanded existing standard IF coverage to validate:

- `totalNumFeatures` persistence
- feature-dimension validation at score time
- empty model transform failure
- invalid `numSamples` handling
- legacy constructor compatibility
- legacy saved model loading without `totalNumFeatures`

### Added / updated test resources

Added snapshot resources for EIF model persistence, including:

- saved extended model metadata
- saved extended tree Avro data
- expected extended tree structure text fixture

Updated standard saved model metadata fixture to include `totalNumFeatures`.

---

### API compatibility

- Existing standard `IsolationForest` API remains available.
- The legacy public 4-arg `IsolationForestModel` constructor is preserved.
- Existing standard model load paths are preserved.

### Persistence compatibility

- Newly saved standard IF models now include `totalNumFeatures`.
- Older saved standard IF models without that metadata still load successfully.

### ONNX compatibility

- No new ONNX behavior is introduced for EIF.
- Existing standard IF ONNX flow remains unchanged.
- EIF is intentionally documented as unsupported for ONNX export.

### Behavior changes to be aware of

This PR introduces a few stricter validation behaviors:

- very small `maxSamples` values that resolve to fewer than 2 samples now fail fast
- transform now rejects empty ensembles
- scoring validates feature vector size when training dimension is known

These are intentional robustness improvements.

---

### Automated test validation

Verified with:

```bash
./gradlew -g /tmp/codex-gradle-home :isolation-forest:test --tests com.linkedin.relevance.isolationforest.IsolationForestModelWriteReadTest --tests com.linkedin.relevance.isolationforest.extended.ExtendedIsolationForestModelWriteReadTest
./gradlew -g /tmp/codex-gradle-home :isolation-forest:test
```

### Benchmark validation in README

The README now includes a detailed benchmark comparison across 13 datasets for:

- standard IF
- EIF with `extensionLevel = 0`
- fully extended EIF
- Liu et al. 2008 paper results
- the reference Python EIF implementation

The headline result is that:

- **StandardIF** remains in strong agreement with the original Liu et al. paper
- **ExtendedIF_max** closely tracks the reference Python EIF implementation across the benchmark suite, with differences generally within a couple of standard errors

### Additional edge-case study (not included in this PR)

In addition to the checked-in unit/integration tests and README benchmark comparison, we also ran a separate detailed hyperparameter / edge-case study outside the PR. This is not part of the checked-in test suite, but it gives extra confidence in the implementation.

**Summary:** 61 / 61 checks passed in ~20s

#### Hyperparameter behavior

- **Extension level (ionosphere, 33D):** AUROC increases monotonically from `0.860` (`ext=0`) to `0.908` (`ext=32`), confirming that higher extension levels capture more complex anomaly structure.
- **Number of trees (breastw):** AUROC is stable at about `0.985+` across `10–200` trees. More trees help slightly, but returns diminish beyond ~50 trees.
- **Sample size (pima):** AUROC ranges from `0.63–0.67` across `32–512` samples. Larger samples are slightly better. Fractional `maxSamples=0.5` behaves correctly.
- **Max features:** Halving the feature subspace (`maxFeatures=0.5`) has minimal effect on standard IF but slightly improves EIF on ionosphere (`0.908 -> 0.913`).
- **Bootstrap:** on/off produces nearly identical results for both IF and EIF.

#### Correctness checks

- **Contamination** only affects the prediction threshold, not the scores themselves. `contamination=0` labels nothing; `contamination=0.35` labels about `35.6%`.
- **Seed reproducibility:** same seed gives bit-identical scores; different seeds produce different scores.
- **Save/load:** Avro round-trip preserves scores exactly (`max diff = 0`).
- **IF vs EIF ext=0:** similar but not identical AUROC (`diff 0.002–0.017`), confirming that they are related but algorithmically distinct.
- **Score distribution:** all scores remain in `[0,1]`, and anomalies consistently score higher than normals.

#### Validation and edge cases

- `extensionLevel > dim - 1` is correctly rejected.
- `maxSamples > n` is correctly rejected; `maxSamples = n` and fractional values behave correctly.
- **1D/2D data:** both IF and EIF handle low-dimensional data well (`AUROC 0.97–0.99`).
- **Constant features:** no degradation; IF and EIF both isolate using the non-constant dimensions.
- **All features constant:** does not crash. IF assigns `0.5` everywhere; EIF assigns `0.291` everywhere.
- **Tiny dataset (`n=3`, `ss=2`):** works and produces identical scores for all points. A single-row dataset is correctly rejected.

<details>
<summary>Raw output from the external edge-case study</summary>

```text
======================================================================
  Isolation Forest / Extended IF — Hyperparameter & Edge Case Tests
======================================================================

=== Test: Extension Level Sweep (ionosphere, 33D) ===
  [PASS] ext_sweep ext=0 — AUROC=0.8600
  [PASS] ext_sweep ext=1 — AUROC=0.8733
  [PASS] ext_sweep ext=4 — AUROC=0.8884
  [PASS] ext_sweep ext=8 — AUROC=0.9027
  [PASS] ext_sweep ext=16 — AUROC=0.9063
  [PASS] ext_sweep ext=32 — AUROC=0.9079
  [PASS] ext_sweep ext=0 != ext=max — ext=0 AUROC=0.8600, ext=max AUROC=0.9079

=== Test: Number of Trees Sweep (breastw) ===
  [PASS] num_trees n=10 — AUROC=0.9851
  [PASS] num_trees n=50 — AUROC=0.9862
  [PASS] num_trees n=100 — AUROC=0.9865
  [PASS] num_trees n=200 — AUROC=0.9880
  [PASS] num_trees monotonic trend — n=10: 0.9851, n=100: 0.9865

=== Test: Sample Size Sweep (pima) ===
  [PASS] sample_size ss=32 — AUROC=0.6411
  [PASS] sample_size ss=64 — AUROC=0.6329
  [PASS] sample_size ss=128 — AUROC=0.6380
  [PASS] sample_size ss=256 — AUROC=0.6569
  [PASS] sample_size ss=512 — AUROC=0.6698
  [PASS] sample_size fractional (0.5) — AUROC=0.6616

=== Test: Max Features / Feature Subspace (ionosphere) ===
  [PASS] max_features IF mf=0.5 — AUROC=0.8438
  [PASS] max_features EIF mf=0.5 — AUROC=0.9134
  [PASS] max_features IF mf=1.0 — AUROC=0.8431
  [PASS] max_features EIF mf=1.0 — AUROC=0.9079

=== Test: Bootstrap (breastw) ===
  [PASS] bootstrap IF bs=false — AUROC=0.9865
  [PASS] bootstrap EIF bs=false — AUROC=0.9832
  [PASS] bootstrap IF bs=true — AUROC=0.9878
  [PASS] bootstrap EIF bs=true — AUROC=0.9825

=== Test: Contamination & Threshold (breastw) ===
  [PASS] contamination=0 no predictions — predicted 0 anomalies (expected 0)
  [PASS] contamination=0.35 fraction — predicted fraction=0.356 (expected ~0.35)
  [PASS] contamination does not affect scores — max score diff=0.00e+00
  [PASS] EIF contamination=0 no predictions — predicted 0 anomalies (expected 0)

=== Test: Seed Reproducibility ===
  [PASS] IF seed reproducibility — max diff=0.00e+00
  [PASS] IF different seeds differ — max diff=0.0412
  [PASS] EIF seed reproducibility — max diff=0.00e+00

=== Test: Save/Load Round-Trip ===
  [PASS] IF save/load scores match — max diff=0.00e+00
  [PASS] EIF save/load scores match — max diff=0.00e+00

=== Test: Standard IF vs EIF ext=0 ===
  [PASS] IF_vs_EIF0 ionosphere — IF=0.8431, EIF_0=0.8600, diff=0.0169
  [PASS] IF_vs_EIF0 breastw — IF=0.9865, EIF_0=0.9881, diff=0.0016
  [PASS] IF_vs_EIF0 pima — IF=0.6569, EIF_0=0.6665, diff=0.0096

=== Test: Score Distribution Properties ===
  [PASS] score_range IF in [0,1] — min=0.3382, max=0.6915
  [PASS] score_separation IF — mean_normal=0.3842, mean_anomaly=0.5741
  [PASS] score_range EIF in [0,1] — min=0.3376, max=0.6724
  [PASS] score_separation EIF — mean_normal=0.3699, mean_anomaly=0.5409

=== Test: Extension Level Validation ===
  [PASS] ext_level > dim-1 rejected — raised: requirement failed: parameter extensionLevel given invalid value 8, but must be
  [PASS] ext_level = dim-1 accepted — ext=7 on 8D data

=== Test: Edge Case — Small Dataset (< default sample_size) ===
  [PASS] small_dataset maxSamples>n rejected — raised: requirement failed: parameter maxSamples given invalid value 256.0 specifying th
  [PASS] small_dataset IF maxSamples=n — n=50, scores range=[0.4007, 0.5988]
  [PASS] small_dataset EIF maxSamples=n — n=50, scores range=[0.3952, 0.5522]
  [PASS] small_dataset IF maxSamples=30 — scores range=[0.4250, 0.5825]
  [PASS] small_dataset IF fractional maxSamples=0.5 — scores range=[0.4288, 0.5900]

=== Test: Edge Case — Low Dimensional (1D, 2D) ===
  [PASS] low_dim IF 1D — AUROC=0.9756
  [PASS] low_dim EIF 1D ext=0 — AUROC=0.9786
  [PASS] low_dim IF 2D — AUROC=0.9934
  [PASS] low_dim EIF 2D ext=1 — AUROC=0.9829

=== Test: Edge Case — Constant Features ===
  [PASS] constant_features IF — AUROC=0.9988
  [PASS] constant_features EIF ext=max — AUROC=0.9980
  [PASS] constant_features EIF ext=0 — AUROC=1.0000

=== Test: Edge Case — All Features Constant ===
  [PASS] all_constant IF no crash — score range=[0.5000, 0.5000], std=0.000000
  [PASS] all_constant EIF no crash — score range=[0.2910, 0.2910], std=0.000000

=== Test: Edge Case — Minimal Dataset ===
  [PASS] single_row IF rejected — raised: requirement failed: parameter maxSamples given invalid value 1.0 specifying the
  [PASS] tiny_dataset IF (n=3, ss=2) — scores=[0.011238790132930549, 0.011238790132930549, 0.011238790132930549]
  [PASS] tiny_dataset EIF (n=3, ss=2) — scores=[0.011238790132930549, 0.011238790132930549, 0.011238790132930549]

======================================================================
  SUMMARY: 61 passed, 0 failed (20.2s)
======================================================================
```

</details>

---

## Why this design

This implementation tries to keep EIF aligned with the rest of the library:

- same Spark ML `Estimator` / `Model` pattern as standard IF
- same thresholding flow for contamination
- same save/load ergonomics
- maximum reuse of shared training and metadata logic

At the same time, it preserves the important algorithmic differences required for EIF correctness:

- hyperplane-based tree structure
- sparse hyperplane persistence
- no retry on degenerate EIF splits
- support for zero-size leaves

---

## Limitations / follow-ups

Not included in this PR:

- ONNX export for `ExtendedIsolationForestModel`

The current ONNX converter assumes axis-aligned tree ensembles, so EIF would require a different export representation or converter strategy.

---

## References

- Hariri, S., Carrasco Kind, M., and Brunner, R. J. **Extended Isolation Forest**.  
  https://arxiv.org/abs/1811.02141

- Reference EIF implementation used for comparison:  
  https://github.com/sahandha/eif

### CODE REVIEW

File: isolation-forest/src/main/scala/com/linkedin/relevance/isolationforest/extended/ExtendedIsolationForestModel.scala [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    val dataWithScore = data.withColumn($(scoreCol), transformUDF(col($(featuresCol))))
+    val dataWithScoresAndPrediction = if (outlierScoreThreshold > 0) {
+      dataWithScore.withColumn(
+        $(predictionCol),
+        (col($(scoreCol)) >= outlierScoreThreshold).cast("double"),
+      )
+    } else {
+      // If threshold is not set, default predictedLabel to 0.0
+      dataWithScore.withColumn($(predictionCol), lit(0.0))
+    }
```
[the remark this answers is not in the document]
Comment (jverbus) [reply isolation-forest/src/main/scala/com/linkedin/relevance/isolationforest/extended/ExtendedIsolationForestModel.scala]: outlierScoreThreshold > 0 sentinel (declined): Copilot noted that threshold=0.0 would be treated as "unset". Technically correct, but this mirrors the existing standard IF pattern identically. A threshold of 0.0 (label everything as outlier) is not a practical use case. Fixing it properly requires changing both IF and EIF together in a separate PR.

File: isolation-forest/src/main/scala/com/linkedin/relevance/isolationforest/core/SharedTrainLogic.scala [hunk 2]
```diff
@@ -219,6 +285,11 @@ private[isolationforest] object SharedTrainLogic extends Logging {
 
         // Shuffle, then slice to limit the data
         val dataForTree = rnd.shuffle(x.toSeq).slice(0, numSamples).toArray
+        if (dataForTree.isEmpty)
+          throw new IllegalStateException(
```
Comment (CrustyAIBot) [inline isolation-forest/src/main/scala/com/linkedin/relevance/isolationforest/core/SharedTrainLogic.scala]: Hard-failing the entire fit on a single empty partition can make training flaky for valid but small `numSamples` settings (e.g., `numSamples=2`), since bagging can still yield zero sampled rows with non-zero probability. Would it be safer to retry sampling/building that tree a bounded number of times before aborting the whole job?

Comment (CrustyAIBot) [reply isolation-forest/src/main/scala/com/linkedin/relevance/isolationforest/core/SharedTrainLogic.scala]: Good point — agreed. With the existing 7σ oversampling, a zero-sample partition should be effectively unreachable for valid settings, so fail-fast with a clear error is reasonable and preferable to masking deeper data/config issues. I’m aligned with treating this as non-blocking for this PR.

**Review by CrustyAIBot — APPROVED** [review]
Deep review complete; looks good to merge. Remaining points are follow-up hardening items, not blockers.

**Review by ganiano — APPROVED** [review]
Thanks for the great work, really appreciate it!
